File: app/src/main/java/com/majeur/psclient/service/ActionQueue.java
package com.majeur.psclient.service;

import android.os.Handler;
import android.os.Looper;

import java.util.LinkedList;
import java.util.Queue;


public class ActionQueue {

    public static final Runnable EMPTY_ACTION = new Runnable() {
        @Override
        public void run() {
        }
    };

    private static class Entry {

        Entry(Runnable action, long delay) {
            this.action = action;
            this.delay = delay;
        }

        private Runnable action;
        private long delay;
    }

    private Handler mHandler;

    private Queue<Entry> mQueue;
    private Runnable mLastAction;
    private boolean mIsLooping;

    private boolean mTurnActionInQueue;

    public ActionQueue(Looper looper) {
        mHandler = new Handler(looper);
        mQueue = new LinkedList<>();
        mIsLooping = false;
    }

    public void clear() {
        stopLoop();
        mTurnActionInQueue = false;
        mLastAction = null;
        mQueue.clear();
    }

    public void setLastAction(Runnable action) {
        mLastAction = action;
    }

    public void enqueueTurnAction(Runnable action) {
        enqueueAction(action);
        if (mTurnActionInQueue)
            loopTo(action);

        mTurnActionInQueue = true;
    }

    public void enqueueAction(Runnable action) {
        enqueue(action, 0);
    }

    public void enqueueMajorAction(Runnable action) {
        enqueue(action, 1500);
    }

    public void enqueueMinorAction(Runnable action) {
        enqueue(action, 750);
    }

    private void enqueue(Runnable action, long delay) {
        mQueue.add(new Entry(action, delay));

        if (!mIsLooping)
            startLoop();
    }

    private void startLoop() {
        mIsLooping = true;
        mHandler.post(mLoopRunnable);
    }

    private void stopLoop() {
        mIsLooping = false;
        mHandler.removeCallbacks(mLoopRunnable);
    }

    private void loopTo(Runnable action) {
        stopLoop();
        while (mQueue.element().action != action)
            mQueue.poll().action.run();
        startLoop();
    }

    private final Runnable mLoopRunnable = new Runnable() {
        @Override
        public void run() {
            Entry entry = mQueue.poll();
            entry.action.run();

            if (!mQueue.isEmpty()) {
                mHandler.postDelayed(this, entry.delay);
            } else {
                mTurnActionInQueue = false;
                mIsLooping = false;
                if (mLastAction != null) {
                    mLastAction.run();
                    mLastAction = null;
                }
            }
        }
    };
}

File: psclient/src/main/java/com/majeur/psclient/service/ActionQueue.kt
package com.majeur.psclient.service

import android.os.Handler
import android.os.Looper

class ActionQueue(looper: Looper) {

    private data class Action(val action: ()->Unit, val delay: Long, val isTurn: Boolean)

    private val handler = Handler(looper)

    private val actions = mutableListOf<Action>()
    private var lastAction: (()->Unit)? = null

    private var isLooping = false

    var shouldLoopToLastTurn = true

    fun clear() {
        stopLoop()
        lastAction = null
        actions.clear()
    }

    fun setLastAction(action: (()->Unit)?) {
        lastAction = action
    }

    fun enqueueTurnAction(action: ()->Unit) {
        val turnActionInQueue = actions.any { it.isTurn }
        enqueue(action, 0, isTurn = true)

        // Only skip to the latest turn if we are watching a live battle ie. when shouldLoopToLastTurn is false.
        // Otherwise, do each turn in the queue one at a time
        if (shouldLoopToLastTurn && turnActionInQueue) {
            loopTo(action)
        }
    }

    fun enqueueAction(action: ()->Unit) {
        enqueue(action, 0, isTurn = false)
    }

    fun enqueueMajorAction(action: ()->Unit) {
        enqueue(action, 1500, isTurn = false)
    }

    fun enqueueMinorAction(action: ()->Unit) {
        enqueue(action, 750, isTurn = false)
    }

    private fun enqueue(action: ()->Unit, delay: Long, isTurn: Boolean) {
        actions.add(Action(action, delay, isTurn))
        if (!isLooping) startLoop()
    }

    fun startLoop() {
        if (actions.isEmpty()) return
        isLooping = true
        handler.post(loopRunnable)
    }

    fun stopLoop() {
        isLooping = false
        handler.removeCallbacks(loopRunnable)
    }

    fun skipToNextTurn() {
        if (actions.isEmpty()) return
        if (actions.any { it.isTurn })
            loopTo(actions.first { it.isTurn }.action)
        else
            loopTo({}) // This will loop to the end
    }

    private fun loopTo(targetAction: ()->Unit) {
        val restartLoop = isLooping
        stopLoop()
        do {
            val action = actions.removeAt(0).action
            action.invoke()
        } while (actions.isNotEmpty() && action != targetAction)
        if (restartLoop) startLoop()
    }

    private val loopRunnable: Runnable = object : Runnable {

        override fun run() {
            val entry = actions.removeAt(0)
            entry.action.invoke()
            if (actions.any()) {
                handler.postDelayed(this, entry.delay)
            } else {
                stopLoop()
                lastAction?.invoke()
                lastAction = null
            }
        }
    }

}
